tests/self-uses-rejected/test_self_uses.py

```diff
@@ -0,0 +1,19 @@
+from dds_ci.testing import ProjectOpener, error
+
+
+def test_self_referential_uses_fails(project_opener: ProjectOpener) -> None:
+    proj = project_opener.open('bad_proj')
+    with error.expect_error_marker_re(r'library-json-cyclic-dependency'):
+        proj.build()
+
+
+def test_self_referential_uses_cycle_fails(project_opener: ProjectOpener) -> None:
+    proj = project_opener.open('bad_proj_large_cycle')
+    with error.expect_error_marker_re(r'library-json-cyclic-dependency'):
+        proj.build()
+
+
+def test_self_referential_uses_for_libs_fails(project_opener: ProjectOpener) -> None:
+    proj = project_opener.open('bad_proj_libs')
+    with error.expect_error_marker_re(r'library-json-cyclic-dependency'):
+        proj.build()
```

Comment:
Since the test projects are very small and simple, you can reduce the number of test files by instead using a `tmp_project` and the methods on `Project` to generate the bad projects in-line, making them easier to diagnose and modify in the future. For example:

```python
def test_foo(tmp_project: Project) -> None:
    # Write a source file:
    tmp_project.write('src/a.cpp', 'int main() {}\n')
    # Assigning these members immediately writes the JSON5 file to the project:
    tmp_project.package_json = { 'name': 'foo', 'version': '1.2.3', 'namespace': 'meow' }
    tmp_project.library_json = { 'name': 'bar', 'uses': ['cat/meow', 'dog/bark'] }
    # ... Do test ...
```